### PR TITLE
Allow "to display" in planet descriptions and spaceport descriptions

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Feature:** The planet descriptions and spaceport descriptions can have "to display" clauses to show or hide them based on conditions.

I swear there is an issue for this somewhere, but I can't find it now, possibly due to sleep deprivation.

## Feature Details

```ruby
planet Wherever
    description `This planet is a very nice place`
       to display
           not "terrible things happened"
    spaceport `This is quite a snazzy spaceport`
       to display
           not "terrible things happened"
    description `Frog People invaded and the planet is now almost devoid of life`
       to display
           has "terrible things happened"
    spaceport `The Frog People spaceport is a dangerous place for humans.`
       to display
           has "terrible things happened"
```

Presently, you can *not* remove a planet's spaceport by changing the conditions. This is because the player conditions are not available in some code that queries the existence of a spaceport. If you disable all "spaceport" blocks, there'll still be a spaceport, but it'll have no text. This could be fixed, but it would require reworking a lot of code, so that improvement belongs in another PR.

## Usage Examples

Fine-grained control over the content of the description and spaceport description, without having to keep track of events.

## Testing Done

1. Put this in data/[mapdisplaytest.txt](https://github.com/endless-sky/endless-sky/files/15503844/mapdisplaytest.txt)
2. Land on Earth to start the test.
3. Depart and land to see nothing changed. This is because the `planet Earth` data has been updated, but the condition to change the descriptions is false.
4. Depart and land to see the new planet and spaceport descriptions. (Frog People invaded Earth.)
5. Depart and land again to see the original planet and spaceport descriptions.

If you don't have a save near Earth handy, this one is on Mars: [Friction Diction~3024-05-09 land on earth to begin the test.txt](https://github.com/endless-sky/endless-sky/files/10475532/Friction.Diction.3024-05-09.land.on.earth.to.begin.the.test.txt)

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

File: source/Planet.h [hunk 1]
```diff
@@ -42,6 +44,10 @@ class Wormhole;
 // might choose it as a source or destination.
 class Planet {
 public:
+	using DescriptionStore = std::vector<std::pair<std::string, std::shared_ptr<ConditionSet>>>;
+	using DescriptionItem = DescriptionStore::value_type;
+
```
Comment (RisingLeaf) [inline source/Planet.h:49]:
```suggestion

public:
```

style

**Comment by UnorderedSigh** [comment]
Question for the coders. I could reduce this PR by about 40 lines if I pass a pointer to a ConditionsStore instead of a reference, because then I could use default arguments. Habitually, I avoid pointers to objects, but that may just be due to silly superstitions on my part.

So, should I pass pointers to ConditionsStores and reduce this PR by 40-ish lines?

**Comment by rovermicrover** [comment]
Related PR that has stalled for a while around adding spaceport and outfitting descriptions.

If I could get some eyes on this to say ya or nay before this is merged.

https://github.com/endless-sky/endless-sky/pull/7839

**Comment by UnorderedSigh** [comment]
Last tested and confirmed to work on Thu, 02 Feb 2023 00:44:38

**Comment by UnorderedSigh** [comment]
Last tested and confirmed to work Sun, 19 Feb 2023 03:26:39 UTC.

**Comment by UnorderedSigh** [comment]
Last tested and confirmed to work at Sat, 04 Mar 2023 17:21:57 UTC.

**Comment by UnorderedSigh** [comment]
Last tested and confirmed to work Sat, 25 Mar 2023 16:39:05 UTC

**Comment by UnorderedSigh** [comment]
Last tested and confirmed to work Sun, 16 Apr 2023 21:24:59 UTC

**Comment by UnorderedSigh** [comment]
Last tested and known to work Fri, 12 May 2023 01:49:07 UTC

**Comment by lumbar527** [comment]
This sounds great and could be used for simplifying stuff like post-bombing Geminus descriptions.

**Comment by Quantumshark** [comment]
It might be better handled with a separate PR, but it would be nice to also have access to `"to display"` in ship and outfit descriptions.

**Comment by UnorderedSigh** [comment]
> It might be better handled with a separate PR, but it would be nice to also have access to "to display" in ship and outfit descriptions.

The PlayerInfo isn't available where that information is used. It would require quite a bit of refactoring, or #8529

**Comment by UnorderedSigh** [comment]
Last tested and known to work Mon, 29 May 2023 16:09:27

**Comment by UnorderedSigh** [comment]
Last tested and known to work Wed, 23 Aug 2023 02:55:05 UTC

**Comment by TomGoodIdea** [comment]
@UnorderedSigh Are you able to update and maintain this PR or should we close it and mark as "orphaned"?

**Comment by UnorderedSigh** [comment]
> @UnorderedSigh Are you able to update and maintain this PR or should we close it and mark as "orphaned"?

Will there ever be any interest in merging this feature? If so, I could update it.

**Comment by TomGoodIdea** [comment]
I don't see anything blocking this, and no one has raised any major objections since the PR was opened.

**Comment by UnorderedSigh** [comment]
I'll update it then.

The easiest way is to add a "Description" class that keeps track of the list of strings and conditions (`vector<pair<ConditionSet,string>>`). We'll have a Description in the Port and a Description in the Planet. The Planet will know how to append the two when relevant. In the future, if people want conditional descriptions elsewhere, they can reuse the Description class to do it.

**Comment by UnorderedSigh** [comment]
Actually... I've encountered the fatal flaw in Endless Sky's naming convention. The very reason why nobody uses capital letters for the first letter of each method. The reason why you prefix the accessors and mutators with get and set.

```
class Description...

class Port
{
    Description &Description();
}

Description Port::Description()
{
    return descripation;
}
```

Inside Port, there is no way to tell the difference between the Description class and the Description method.

**Comment by UnorderedSigh** [comment]
Last tested and known to work Thu, 30 May 2024 18:04:15 UTC

File: source/Paragraphs.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#include "ConditionSet.h"
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+class ConditionsStore;
+class DataNode;
+
+class Paragraphs {
```
Comment (RisingLeaf) [inline source/Paragraphs.h:28]:
```suggestion


class Paragraphs {
```
Need a comment on purpose of class

File: source/Paragraphs.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ConditionSet.h"
+#include "ConditionsStore.h"
+#include "DataNode.h"
+#include "Paragraphs.h"
```
Comment (RisingLeaf) [inline source/Paragraphs.cpp:19]:
```suggestion
#include "Paragraphs.h"

#include "ConditionSet.h"
#include "ConditionsStore.h"
#include "DataNode.h"
```

File: source/Paragraphs.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+using namespace std;
+
```
Comment (RisingLeaf) [inline source/Paragraphs.cpp:22]:
```suggestion



```

File: source/Paragraphs.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+bool Paragraphs::IsEmpty() const
+{
+	return text.empty();
+}
+
+
+
+bool Paragraphs::IsEmptyFor(const ConditionsStore &vars) const
+{
+	for(auto &varsText : text)
```
Comment (RisingLeaf) [inline source/Paragraphs.cpp:52]:
```suggestion
	for(const auto &varsText : text)
```
?

File: source/Paragraphs.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+			return false;
+	return true;
+}
+
+
+
+string Paragraphs::ToString(const ConditionsStore &vars) const
+{
+	string result;
+	for(auto &varsText : text)
```
Comment (RisingLeaf) [inline source/Paragraphs.cpp:63]:
```suggestion
	for(const auto &varsText : text)
```
Should work, right?

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/Paragraphs.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef PARAGRAPHS_H_
+#define PARAGRAPHS_H_
+
+#include "ConditionSet.h"
+
+#include <algorithm>
```
Comment (TomGoodIdea) [inline source/Paragraphs.h:21]: What do we need `<algorithm>` for?
```suggestion
```

File: source/Paragraphs.h [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+	bool IsEmptyFor(const ConditionsStore &vars) const;
+
+	// Concatinate all lines which match these vars.
+	std::string ToString(const ConditionsStore &vars) const;
+
+	// Iterate over all text. Needed to support PrintData.
+	// These must use standard naming conventions (begin, end) for compatibility with range-based for loops.
+	ConstIterator begin() const;
+	ConstIterator end() const;
+
```
Comment (TomGoodIdea) [inline source/Paragraphs.h:54]:
```suggestion


```

File: source/Paragraphs.h [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+private:
+	ConditionalText text;
+};
+
```
Comment (TomGoodIdea) [inline source/Paragraphs.h:58]:
```suggestion



```

File: source/Paragraphs.cpp (hunk 6 above)
Comment (UnorderedSigh) [reply source/Paragraphs.cpp:63]: It is redundant. The method is const, so any iterator over a member is const (unless it is a mutable member, but that is naughty, and we don't do naughty things).

File: source/Paragraphs.h (hunk 7 above)
Comment (UnorderedSigh) [reply source/Paragraphs.h]: That is one of the ways you can get the "pair" class. The "proper" way is to use utility, not algorithm. I've updated the code.

File: source/Paragraphs.h (hunk 2 above)
Comment (UnorderedSigh) [reply source/Paragraphs.h:32]: How's this one?

> // Stores a list of description paragraphs, and a condition under which each should be shown.
> // See the planet and spaceport description code for examples.

File: source/Paragraphs.cpp (hunk 6 above)
Comment (UnorderedSigh) [reply source/Paragraphs.cpp]: I committed the suggestion anyway. Hopefully nobody will ask me to change it back.

File: source/Paragraphs.h (hunk 7 above)
Comment (UnorderedSigh) [reply source/Paragraphs.h]: pair-header-utility

File: source/Paragraphs.h [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	// Discard all description lines.
+	void Clear();
+
+	// Is this object totally void of all information?
+	bool IsEmpty() const;
+
+	// Are there any lines which match these vars?
+	bool IsEmptyFor(const ConditionsStore &vars) const;
+
+	// Concatinate all lines which match these vars.
```
Comment (TomGoodIdea) [inline source/Paragraphs.h:51]:
```suggestion
	// Concatenate all lines which match these vars.
```

**Review by TomGoodIdea — APPROVED** [review]

**Review by RisingLeaf — APPROVED** [review]

**Review by tibetiroka — APPROVED** [review]

**Pull request merged**